database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -221,16 +231,21 @@ public void onDataChange(DataSnapshot snapshot) {
                     // We don't already know about this data, add it
                     mDataSnapshots.add(index, snapshot);
                     notifyChangeEventListeners(ChangeEventListener.EventType.ADDED, snapshot, index);
+
+                    mQueuedAdditions.remove(key);
+                    if (mQueuedAdditions.isEmpty()) notifyListenersOnDataChanged();
                 } else {
                     // We already know about this data, just update it
                     updateData(index, snapshot);
                     notifyChangeEventListeners(ChangeEventListener.EventType.CHANGED, snapshot, index);
+                    notifyListenersOnDataChanged();
```

Comment:
I can't think of a better way to do this without having some sort of looper and/or timer. @samtstern Do you have any ideas?